lineapy/graph_reader/graph_refactorer.py

```diff
@@ -0,0 +1,507 @@
+import itertools
+import logging
+from collections import OrderedDict
+from dataclasses import dataclass
+from enum import Enum
+from typing import Any, Dict, List, Optional, Set
+
+import networkx as nx
+
+from lineapy.api.api_classes import LineaArtifact
+from lineapy.data.graph import Graph
+from lineapy.data.types import LineaID, Node
+from lineapy.db.db import RelationalLineaDB
+from lineapy.graph_reader.program_slice import (
+    get_slice_graph,
+    get_source_code_from_graph,
+)
+from lineapy.utils.logging_config import configure_logging
+
+logger = logging.getLogger(__name__)
+configure_logging()
+
+
+class GraphSegmentType(Enum):
+    """
+    GraphSegment type to identify the purpose of the graph segment
+    - ARTIFACT : the segment returns an artifact
+    - HELPER : the segment returns variables used in multiple artifacts
+    """
+
+    ARTIFACT = 1
+    COMMON_VARIABLE = 2
+
+
+@dataclass
+class GraphSegment:
+    """
+    This class contains information required to wrap a lineapy Graph object as a function
+    """
+
+    artifact_name: str
+    artifact_safename: str
+    graph_segment: Graph
+    segment_type: GraphSegmentType
+    all_variables: Set[str]
+    input_variables: Set[str]
+    return_variables: List[str]
+
+    def __init__(
+        self,
+        artifact_name,
+        graph_segment,
+        segment_type,
+        all_variables,
+        input_variables,
+        return_variables,
+    ) -> None:
+        self.artifact_name = artifact_name
+        self.artifact_safename = artifact_name.replace(" ", "")
+        self.graph_segment = graph_segment
+        self.segment_type = segment_type
+        self.all_variables = all_variables
+        self.input_variables = input_variables
+        self.return_variables = return_variables
+
+    def get_function_definition(self, indentation=4) -> str:
+        """
+        Return a codeblock to define the function
+        """
+        indentation_block = " " * indentation
+        artifact_code = get_source_code_from_graph(
+            self.graph_segment
+        ).__str__()
+        artifact_codeblock = "\n".join(
+            [
+                f"{indentation_block}{line}"
+                for line in artifact_code.split("\n")
+            ]
+        )
+        artifact_name = self.artifact_safename
+        args_string = ", ".join([v for v in self.input_variables])
+        return_string = ", ".join([v for v in self.return_variables])
+
+        return f"def get_{artifact_name}({args_string}):\n{artifact_codeblock}\n{indentation_block}return {return_string}"
+
+    def get_function_call_block(self, keep_lineapy_save=False) -> str:
```

Comment:
Can we add `indentation: int = 0` arg to support easier code embedding.

Reply:
done